cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,299 @@
+            thrust::get<1>(rank_count1) + thrust::get<1>(rank_count2)};
+      },
+      [] __device__(MinCount minrank_count) {  // min+(count-1)/2
+        return static_cast<double>(thrust::get<0>(minrank_count)) +
+              (static_cast<double>(thrust::get<1>(minrank_count)) - 1) / 2.0;
+      },
+      stream);
+}
+
+} // anonymous namespace
+
+std::unique_ptr<column> rank(column_view const &input, rank_method method,
+                            order column_order, include_nulls _include_nulls,
+                            null_order null_precedence,
+                            bool percentage,
+                            rmm::mr::device_memory_resource *mr,
+                            cudaStream_t stream = 0) {
+  auto const input_size = input.size();
+
+  std::unique_ptr<column> rank_column = [&] {
+    // na_option=keep assign NA to NA values
+    if (_include_nulls == include_nulls::NO)
+      return make_numeric_column(data_type(FLOAT64), input_size, copy_bitmask(input, stream, mr), input.null_count(), stream, mr);
+    else
+      return make_numeric_column(data_type(FLOAT64), input_size, mask_state::UNALLOCATED, stream, mr);
+  }();
+  auto rank_mutable_view = rank_column->mutable_view();
+  auto rank_data = rank_mutable_view.data<double>();
+
+  std::unique_ptr<column> sorted_order = (method == rank_method::FIRST)
+          ? detail::stable_sorted_order(table_view{{input}}, {column_order}, {null_precedence}, mr, stream)
+          : detail::sorted_order       (table_view{{input}}, {column_order}, {null_precedence}, mr, stream);
+  column_view sorted_order_view = sorted_order->view();
+
+  // All equal values have same rank and rank always increases by 1 between groups
+  // acts as key for min, max, average to denote equal value groups
+  rmm::device_vector<size_type> dense_rank_sorted = [&] {
+    if (method != rank_method::FIRST)
+      return sorted_dense_rank(input, sorted_order_view, stream);
+    else
+      return rmm::device_vector<size_type>();
+  }();
+
+  switch (method) {
+    case rank_method::FIRST:
+      rank_first  (sorted_order_view, rank_mutable_view, stream);
+      break;
+    case rank_method::DENSE:
+      rank_dense  (dense_rank_sorted, sorted_order_view, rank_mutable_view, stream);
+      break;
+    case rank_method::MIN:
+      rank_min    (dense_rank_sorted, sorted_order_view, rank_mutable_view, stream);
+      break;
+    case rank_method::MAX:
+      rank_max    (dense_rank_sorted, sorted_order_view, rank_mutable_view, stream);
+      break;
+    case rank_method::AVERAGE:
+      rank_average(dense_rank_sorted, sorted_order_view, rank_mutable_view, stream);
+      break;
+    default:
+      CUDF_FAIL("Unexpected rank_method for rank()");
+  }
+
+  // pct inplace transform
+  if (percentage) {
+    size_type const count = (_include_nulls == include_nulls::NO)
+                                        ? input_size - input.null_count()
+                                        : input_size;
```

Comment:
Are you using clang-format on this file? Your indenting seems to be ad hoc. Here is seems custom, most places it seems 2 spaces per indent, some places (line 199-214, 279-284) it seems to be 4 spaces? You should try to be consistent with your indenting.

Reply:
I use clang-format on sections of code, not entire file.
I am not sure why auto indent messes it up. vs code editor seems to use 2 spaces(on pressing enter), clang-format uses 4 spaces. 
in clang-format, code folding of single statement to multiple line uses 4 spaces. Statements of children block are indented using 2 spaces. 

Thanks. This helps a lot.

Reply:
formatted whole file with clang-format and reformatted few lines for readability.